### PR TITLE
Add support for virtio-mmio based transport

### PR DESCRIPTION
**Pull request by rbradford**
This PR still needs some polish as there are various clippy warnings due to conditional compilation. And it also lacks some new integration tests that will require us to build an alternative version of the binary.

### CODE REVIEW

**Comment by rbradford** [comment]
The travis failure is due to clippy which I will rectify tomorrow.

**Comment by rbradford** [comment]
(The non top-level features are "pci_support" as you can't have a feature that shares it's name with a crate you use and "mmio_support" is named to be similar)

**Comment by sameo** [comment]
> @rbradford very nice work as always!
> I have one comment about the tests. I think it would be nice to duplicate with direct kernel boot all the tests that might be meaningful to try with `mmio`.

Agreed. Essentially, all virtio specific tests should be run with both transports. I also agree that this can be tackled through a follow up PR. This PR is excellent, LGTM.

File: vmm/src/device_manager.rs [hunk 1]
```diff
@@ -934,6 +964,63 @@ impl DeviceManager {
         Ok(())
     }
 
+    #[allow(clippy::too_many_arguments)]
```
[the remark this answers is not in the document]
Comment (rbradford) [reply vmm/src/device_manager.rs]: I have some thoughts but that would be an even bigger refactor. The 7 limit is somewhat arbitrary and wrapping data into a struct just to keep clippy happy has lowish utility.

File: vmm/src/device_manager.rs [hunk 2]
```diff
@@ -403,31 +408,56 @@ impl DeviceManager {
...
-            for device in virtio_devices {
-                DeviceManager::add_virtio_pci_device(
-                    device,
-                    vm_info.memory,
-                    allocator,
-                    vm_info.vm_fd,
-                    &mut pci,
-                    &mut buses,
-                    &interrupt_info,
+        if cfg!(feature = "pci_support") {
```
[the remark this answers is not in the document]
Comment (rbradford) [reply vmm/src/device_manager.rs]: Great question. The belts and braces if/if else and block conditional code is so handle the case where mmio and pci are built in. I don't expect this to occur for users but this is a feature combination clippy uses.

File: src/main.rs [hunk 3]
```diff
@@ -1348,17 +1353,17 @@ mod tests {
...
     }
 
-    #[test]
+    #[cfg_attr(not(feature = "mmio"), test)]
     fn test_virtio_fs_dax_on_cache_size_1_gib() {
         test_virtio_fs(true, Some(0x4000_0000), "always")
     }
 
-    #[test]
+    #[cfg_attr(not(feature = "mmio"), test)]
```
[the remark this answers is not in the document]
Comment (rbradford) [reply src/main.rs]: This is disabled because I did test it and it does fail.

**Comment by rbradford** [comment]
@sameo I have some thoughts about how to handle testing code for both transports that I will explore in a follow-up PR.

**Pull request merged**